### PR TITLE
Replace notifications with user journey events (GSI-669)

### PR DESCRIPTION
**Pull request by TheByronHimes**
Removes any references to notifications and the notification-publishing machinery (e.g. NotificationEmitter). These are replaced by a vanilla event publisher that publishes the same event schema under different event types, as configured. 

### Changes in the main package
- Major version number bumped from 1 to 2
- Config updated to remove notification-related and data steward email settings, but add access request event settings
- Template updates, which pulled in the `uv` dependency
- The outbound notification emitter adapter was replaced with an event publisher translator (EventPubTranslatorConfig/EventPubTranslator)
- The core module `notifications.py` was removed entirely because that information is now in the NOS.
- The change in event publishing can be summarized with:
  - old: publish 2 Notification events when an access request is created and 2 more when it is allowed or denied.
  - new: publish 1 AccessRequestDetails event when a request is created, and 1 more when it is allowed or denied.
    - Each status corresponds to one event type, which resides in configuration.
 
### Changes in testing
- `test_api.py`:
  - The `assert_recorded_events` function was removed from test_api.py because it's no longer needed.
  - The `NotificationPayload` NamedTuple was deleted from test_api.py as well.
  - Deleted the block in both `test_create_access_request` and `test_patch_access_request` that verified the recorded events, replaced with a straight content-check of the AccessRequestDetails event and metadata (key, topic, type).
- `test_event_pub.py`:
  - Replaced `test_sending_a_notification` with `test_publishing_an_event`, which makes sure that AccessRequest objects are translated into events with the correct payload and metadata.
- `test_repository.py`:
  - Replaced `NotificationRecord` with `AccessRequestDetails` and moved it to tests.fixtures.__init__.py.
  - Replaced the `NotificationEmitterDummy` with a comparable `EventPublisherDummy`.
    - This holds a dict where the keys are the tuple of (user_id, dataset_id) and the value is the status of the request.
  - Modified the tests `test_can_create_request` and `test_set_status_to_allowed` to check that the correct event was called and that the request was passed to the event publisher dummy correctly.

### CODE REVIEW

File: tests/test_event_pub.py [hunk 1]
```diff
@@ -54,27 +60,49 @@ async def _publish_validated(
...
+        expected_type = dummy_config.access_request_created_type
+        await event_publisher.publish_request_created(request=request)
+    elif status == "allowed":
+        expected_type = dummy_config.access_request_allowed_type
+        await event_publisher.publish_request_allowed(request=request)
+    elif status == "denied":
+        expected_type = dummy_config.access_request_denied_type
+        await event_publisher.publish_request_denied(request=request)
+    else:
+        raise RuntimeError("Double-check test parameters")
```
Comment (Cito) [inline tests/test_event_pub.py:101]:
```suggestion
    publish_method = getattr(event_publisher, f"publish_request_{status}")
    await publish_method(request=request)

    expected_topic = dummy_config.access_request_events_topic
    expected_type = getattr(dummy_config, f"access_request_{status}_type")
    expected_payload = AccessRequestDetails(
        user_id=request.user_id, dataset_id=request.dataset_id
    ).model_dump()
```

this is a bit simpler, and you also get an error if there is a wrong test parameter

File: tests/test_event_pub.py [hunk 2]
```diff
@@ -54,27 +60,49 @@ async def _publish_validated(
...
-        "email_cc": [],
-        "email_bcc": [],
-        "subject": "Some subject",
-        "plaintext_body": "Some text",
-        "topic": "dummy_topic",
-        "type": "dummy_type",
-        "key": "someone@somewhere.org",
+        "type": expected_type,
+        "key": request.user_id,
+        "topic": dummy_config.access_request_events_topic,
```
Comment (Cito) [inline tests/test_event_pub.py:106]:
```suggestion
        "topic": expected_topic,
```

File: tests/test_repository.py [hunk 3]
```diff
@@ -190,39 +190,52 @@ async def update(self, dto: AccessRequest) -> None:
...
-    """Dummy notification emitter for testing."""
-
-    notifications: dict[str, NotificationRecord]
+    events: dict[AccessRequestDetails, str]
 
     def reset(self) -> None:
-        """Reset the recorded notification."""
-        self.notifications = {}
+        """Reset the recorded events."""
+        self.events = {}
```
Comment (Cito) [inline tests/test_repository.py:200]: Maybe better to use a list? Then you would also catch if the same events are sent out twice.

**Review by Cito — APPROVED** [review]
Looks good. Just a few small suggestions.

File: src/ars/adapters/outbound/event_pub.py [hunk 4]
```diff
@@ -22,54 +22,82 @@
...
+    access_request_events_topic: str = Field(
+        default=...,
+        description="The topic used for events related to access requests.",
+        examples=["access_requests"],
     )
-    notification_event_type: str = Field(
-        ...,
-        description=("The type used for notification events."),
-        examples=["notification"],
+    access_request_created_type: str = Field(
```
[the remark this answers is not in the document]
Comment (TheByronHimes) [reply src/ars/adapters/outbound/event_pub.py:39]: That is a good point that I didn't catch. I think it's still 'access_request_created_type' in the NOS, so I'll need to change that there too.

File: tests/test_api.py [hunk 5]
```diff
@@ -115,9 +83,9 @@ async def test_create_access_request(
 ):
     """Test that an active user can create an access request."""
     kafka = joint_fixture.kafka
-    topic = joint_fixture.config.notification_event_topic
+    topic = joint_fixture.config.access_request_events_topic
     async with kafka.record_events(in_topic=topic):
-        pass  # skip previous notifications
+        pass  # skip previous events
```
[the remark this answers is not in the document]
Comment (TheByronHimes) [reply tests/test_api.py:88]: I will do that. I have some ideas, but wasn't able to reach a conclusion just now.

File: tests/test_repository.py (hunk 3 above)
Comment (TheByronHimes) [reply tests/test_repository.py]: Changes are in [6ba34d3](https://github.com/ghga-de/access-request-service/pull/14/commits/6ba34d329a02de775cd90731239135b05e8bed41), good idea.

Comment (Cito) [reply tests/test_repository.py]: Actually I meant just a plain list, not a (default)dict of lists.

But if you prefer the defaultdict, then you don't need to catch KeyErrors.

**Review by Cito — DISMISSED** [review]

File: tests/test_repository.py (hunk 3 above)
Comment (TheByronHimes) [reply tests/test_repository.py]: Please see [6ba34d3](https://github.com/ghga-de/access-request-service/pull/14/commits/6ba34d329a02de775cd90731239135b05e8bed41), where I switched to a list and store a more direct mock event

**Review by Cito — DISMISSED** [review]

**Review by Cito — APPROVED** [review]

**Pull request merged**